Mathlib/Analysis/SpecialFunctions/BinEntropy.lean

```diff
@@ -0,0 +1,508 @@
+    have logConcave := (strictConcaveOn_log_Ioi.right
+      (x := 1/p) (y := 1/(1-p))) (a := p) (b := 1-p)
+      invppos sub1pinvpos (by norm_num; linarith) (by positivity)
+      (by linarith) (by norm_num)
+    have : p • (1 / p) + (1 - p) • (1 / (1 - p)) = 2 := by field_simp; norm_num
+    rw [this] at logConcave
+    have := calc -(p * log p) - (1 - p) * log (1 - p)
+      _ = p * (-log p) + (1 - p) * (-log (1 - p)) := by ring
+      _ = p * log (1/p) + (1 - p) * log (1 / (1 - p)) := by rw [← log_inv]; norm_num
+    rw [this]
+    exact logConcave
+
+lemma binaryEntropy_lt_log2_of_gt_half {p : ℝ} (h : 1/2 < p) (h2 : p ≤ 1) :
+    binaryEntropy p < log 2 := by
+  rw [← binaryEntropy_one_sub]
+  exact binaryEntropy_lt_log2_of_lt_one_half (by linarith) (by linarith)
+
+lemma binaryEntropy_eq_log2_iff_eq_half {p : ℝ} (pge0 : 0 ≤ p) (ple1 : p ≤ 1) :
+    binaryEntropy p = log 2 ↔ p = 1/2 := by
+  constructor <;> intro h
+  · by_cases h' : p < 1/2
+    · linarith [binaryEntropy_lt_log2_of_lt_one_half pge0 h']
+    · by_cases pgthalf : 1/2 < p
+      · linarith [binaryEntropy_lt_log2_of_gt_half pgthalf ple1]
+      · linarith
+  · simp only [one_div, binaryEntropy_onehalf, h]
+
+lemma binaryEntropy_le_log2 {p : ℝ} (pge0 : 0 ≤ p) (ple1 : p ≤ 1) :
+    binaryEntropy p ≤ log 2 := by
+  by_cases hh: p = 1/2
+  · simp only [one_div, binaryEntropy_onehalf, le_refl, hh]
+  · by_cases gg: binaryEntropy p = log 2
+    · simp only [le_refl, gg]
+    · by_cases hhh: p < 1/2
+      · linarith [binaryEntropy_lt_log2_of_lt_one_half pge0 hhh]
+      · have : 1/2 < p := by
+          refine Ne.lt_of_le ?_ ?_
+          · intro
+            simp_all only [not_lt, not_true_eq_false]
+          · simp_all only [one_div, not_lt]
+        exact (binaryEntropy_lt_log2_of_gt_half this ple1).le
+
+/-- The q-ary entropy function is continuous everywhere.
+This is due to definition of `Real.log` for negative numbers. -/
+@[fun_prop] lemma qaryEntropy_continuous {q : ℕ} : Continuous (qaryEntropy q) := by
+  refine Continuous.add ?_ (Continuous.neg ?_)
+  · exact Continuous.sub (by fun_prop) continuous_mul_log
+  · exact Continuous.comp continuous_mul_log (continuous_sub_left 1)
+
+/-- Binary entropy is continuous everywhere.
+This is due to definition of `Real.log` for negative numbers. -/
+@[fun_prop] lemma binaryEntropy_continuous : Continuous binaryEntropy := qaryEntropy_continuous
+
+/-! ### Derivatives of binary entropy function -/
+
+lemma differentiableAt_binaryEntropy {x : ℝ} (xne0: x ≠ 0) (gne1 : x ≠ 1) :
+    DifferentiableAt ℝ binaryEntropy x := by
+  simp only [binaryEntropy_eq]
+  refine DifferentiableAt.sub (DifferentiableAt.mul (by fun_prop) ?_)
+      (DifferentiableAt.mul (by fun_prop)
+      (DifferentiableAt.log (by fun_prop) (sub_ne_zero.mpr gne1.symm)))
+  exact DifferentiableAt.log differentiableAt_id' xne0
+
+lemma differentiableAt_binaryEntropy_iff_ne_zero_one (x : ℝ) :
+    DifferentiableAt ℝ binaryEntropy x ↔ x ≠ 0 ∧ x ≠ 1 := by
```

Comment:
```suggestion
    DifferentiableAt ℝ binaryEntropy x ↔ x ≠ 0 ∧ x ≠ 1 := by
```

Reply:
Is that a change?

Reply:
Whoop sorry, not sure what happened there